server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -120,74 +296,357 @@ public void collect(int doc, long bucket) throws IOException {
+                bucketOrds = new LongKeyedBucketOrds.FromMany(context.bigArrays());
+                for (long owningBucketOrd = 0; owningBucketOrd <= oldOrds.maxOwningBucketOrd(); owningBucketOrd++) {
+                    LongKeyedBucketOrds.BucketOrdsEnum ordsEnum = oldOrds.ordsEnum(owningBucketOrd);
+                    Rounding.Prepared preparedRounding = preparedRoundings[roundingIndexFor(owningBucketOrd)];
+                    while (ordsEnum.next()) {
+                        long oldKey = ordsEnum.value();
+                        long newKey = preparedRounding.round(oldKey);
+                        long newBucketOrd = bucketOrds.add(owningBucketOrd, newKey);
+                        mergeMap[(int) ordsEnum.ord()] = newBucketOrd >= 0 ? newBucketOrd : -1 - newBucketOrd;
                     }
-                    bucketOrds = newBucketOrds;
+                    liveBucketCountUnderestimate = context.bigArrays().grow(liveBucketCountUnderestimate, owningBucketOrd + 1);
+                    liveBucketCountUnderestimate.set(owningBucketOrd, Math.toIntExact(bucketOrds.bucketsInOrd(owningBucketOrd)));
                 }
+                merge(mergeMap, bucketOrds.size());
             }
-        };
-    }
+        }
 
-    @Override
-    public InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException {
-        return buildAggregationsForVariableBuckets(owningBucketOrds, bucketOrds,
-                (bucketValue, docCount, subAggregationResults) ->
-                    new InternalAutoDateHistogram.Bucket(bucketValue, docCount, formatter, subAggregationResults),
-                buckets -> {
-                    // the contract of the histogram aggregation is that shards must return
-                    // buckets ordered by key in ascending order
-                    CollectionUtil.introSort(buckets, BucketOrder.key(true).comparator());
-
-                    // value source will be null for unmapped fields
-                    InternalAutoDateHistogram.BucketInfo emptyBucketInfo = new InternalAutoDateHistogram.BucketInfo(roundingInfos,
-                            roundingIdx, buildEmptySubAggregations());
-
-                    return new InternalAutoDateHistogram(name, buckets, targetBuckets, emptyBucketInfo, formatter, metadata(), 1);
-                });
-    }
+        @Override
+        public InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException {
+            /*
+             * Rebucket before building the aggregation to build as small as result
+             * as possible.
+             *
+             * TODO it'd be faster if we could apply the merging on the fly as we
+             * replay the hits and build the buckets. How much faster is not clear,
+             * but it does have the advantage of only touching the buckets that we
+             * want to collect.
+             */
+            rebucket();
```

Comment:
Hmm, we might need to guard here in the situation where the agg doesn't actually generate any buckets (none of the collected docs end up having the field, no collected docs, etc), and return an empty agg instead.

Or alternatively, put the guard in `rebucket()`.

I think `BucketsAggregator#mergeBuckets()` would be fine since doc counts would also be empty... but I'm less confident about `MergingBucketsDeferringCollector#mergeBuckets()`.  Might be ok though, only skimmed it :)  We've ran into this kind of issue in the past though where buildAgg broke when the agg runs and is effectively "unmapped" after running, so thought I'd mention :)